### DISCUSSION TITLE
converting pydantic BaseSettings into pythonic resources

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#15210] [Q&A]
I was trying to take a stab at the migration to pythonic resources (<https://docs.dagster.io/guides/dagster/migrating-to-pythonic-resources-and-config|see here>), but I'm getting a little confused about the best way to deal with the following case: We have quite a lot of existing configuration that is based on pydantic `BaseModel` and/or `BaseSettings` (the structures here are quite nested and complex). As far as I can tell, dagster `ConfigurableResource` should be very similar to this, and likely based on the same pydantic base classes.

Now, I would like to turn few of these top-level configs into dagster resources, but I'm not quite sure how to go about that.

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1689026299751459?thread_ts=1689026299.751459&cid=C01U954MEER).*

**Comment by dpeng817** [comment]
Pulling inspiration from an internal thread with @benpankow :
```python
from pydantic import create_model, BaseSettings


def basesettings_to_resource(bs: BaseSettings) -> Type[ConfigurableResource]:
    return create_model(
        dc.__name__,
        **{k: (v.type_, v.field_info) for k, v in bs.__pydantic_model__.__fields__.items()},
        __base__=ConfigurableResource,
    )
```
that way you can do something like one of the following:
```python
@basesettings_to_resource
class MySettings(BaseSettings):
    my_int: int
    my_str: str = "default"
```

**Comment by mjclarke94** [comment]
Just for reference, we took a slightly different approach in our implementation of this, which I outlined [here](https://github.com/dagster-io/dagster/issues/14571#issuecomment-1576365675). Basically, creating a dummy subclass which inherits both from the Dagster config base class and also your own BaseModel derived class seems to work. 

It's slightly heavy handed, but works well when you're importing the definitions from an external library (i.e. aren't in a position to put a decorator in the definition), or want to keep the Config in a library which doesn't have dagster as a dependency.
